internal/pkg/term/syncbuffer/termprinter.go

```diff
@@ -0,0 +1,176 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package syncbuffer provides a goroutine safe bytes.Buffer as well printing functionality to the terminal.
+package syncbuffer
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"strings"
+
+	"github.com/aws/copilot-cli/internal/pkg/term/cursor"
+	"golang.org/x/term"
+)
+
+// printAllLinesInBuf represents to print the entire contents in the buffer.
+const (
+	printAllLinesInBuf = -1
+)
+
+// FileWriter is the interface to write to a file.
+type FileWriter interface {
+	io.Writer
+	Fd() uintptr
+}
+
+// LabeledTermPrinter is a printer to display label and logs to the terminal.
+type LabeledTermPrinter struct {
+	term             FileWriter           // term writes logs to the terminal FileWriter.
+	buffers          []*LabeledSyncBuffer // buffers stores logs before writing to the terminal.
+	numLines         int                  // number of lines that has to be written from each buffer.
+	padding          int                  // Leading spaces before rendering to terminal.
+	prevWrittenLines int                  // number of lines written from all the buffers.
+
+	// override in unit tests.
+	terminalWidth func(fw FileWriter) (int, error)
+}
+
+// LabeledTermPrinterOption is a type alias to configure LabeledTermPrinter.
+type LabeledTermPrinterOption func(ltp *LabeledTermPrinter)
+
+// NewLabeledTermPrinter returns a LabeledTermPrinter that can print to the terminal filewriter from buffers.
+func NewLabeledTermPrinter(fw FileWriter, bufs []*LabeledSyncBuffer, opts ...LabeledTermPrinterOption) *LabeledTermPrinter {
+	ltp := &LabeledTermPrinter{
+		term:          fw,
+		buffers:       bufs,
+		numLines:      printAllLinesInBuf, // By default set numlines to -1 to print all from buffers.
+		terminalWidth: terminalWidth,
+	}
+	for _, opt := range opts {
+		opt(ltp)
+	}
+	return ltp
+}
+
+// WithNumLines sets the numlines of LabeledTermPrinter.
+func WithNumLines(n int) LabeledTermPrinterOption {
+	return func(ltp *LabeledTermPrinter) {
+		ltp.numLines = n
+	}
+}
+
+// WithPadding sets the padding of LabeledTermPrinter.
+func WithPadding(n int) LabeledTermPrinterOption {
+	return func(ltp *LabeledTermPrinter) {
+		ltp.padding = n
+	}
+}
+
+// IsDone returns true if all the buffers are done.
+func (ltp *LabeledTermPrinter) IsDone() bool {
+	for _, buf := range ltp.buffers {
+		if !buf.IsDone() {
+			return false
+		}
+	}
+	return true
+}
+
+// Print prints the label and the last N lines of logs from each buffer
+// to the LabeledTermPrinter fileWriter and erases the previous output.
+// If numLines is -1 then print all the values from buffers.
+func (ltp *LabeledTermPrinter) Print() error {
+	if ltp.numLines == printAllLinesInBuf {
+		ltp.printAll()
+		return nil
+	}
+	if ltp.prevWrittenLines > 0 {
+		cursor.EraseLinesAbove(ltp.term, ltp.prevWrittenLines)
+	}
```

Comment:
does it cause errors to remove the `if` and just call `cursor.EraseLinesAbove(ltp.term, 0)` if `prevWrittenLines` is 0?

Reply:
It will an error. If i remove if block then `EraseLinesAbove` will try to erase the [current line](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/term/cursor/cursor.go#L98) as well.